addons/pos_mercado_pago/controllers/main.py

```diff
@@ -0,0 +1,92 @@
+        external_reference = data.get('additional_info', {}).get('external_reference')
+
+        if not external_reference or not self.validate_external_reference(external_reference):
+            _logger.debug('POST message received with no or malformed "external_reference" key')
+            return None
+
+        session_id, payment_method_id = external_reference.split('_')
+
+        pos_session_sudo = request.env['pos.session'].sudo().browse(int(session_id))
+        if not pos_session_sudo or pos_session_sudo.state != 'opened':
+            _logger.error("pos_session_sudo invalid")
+            # This error is not related with Mercado Pago, simply acknowledge Mercado Pago message
+            return request.make_json_response('[success]')
+
+        payment_method_sudo = pos_session_sudo.config_id.payment_method_ids.filtered(lambda p: p.id == int(payment_method_id))
+        if not payment_method_sudo or payment_method_sudo.use_payment_terminal != 'mercado_pago':
+            _logger.error("wrong payment terminal")
+            # This error is not related with Mercado Pago, simply acknowledge Mercado Pago message
+            return request.make_json_response('[success]')
+
+        # We have to check if this comes from Mercado Pago with the secret key
+        secret_key = payment_method_sudo.mp_webhook_secret_key
+        sig = x_signature.split(',')
+        ts = sig[0].split('=')[1]
+        v1 = sig[1].split('=')[1]
+        signed_template = f"id:{data['id']};request-id:{x_request_id};ts:{ts};"
+        # cyphed_signature = binascii.hexlify(hmac.new(secret_key.encode(), signed_template.encode(), hashlib.sha256).digest())
+        cyphed_signature = hmac.new(secret_key.encode(), signed_template.encode(), hashlib.sha256).hexdigest()
+        # if cyphed_signature.decode() != v1:
+        if not hmac.compare_digest(cyphed_signature, v1):
+            _logger.error('webhook authenticating failure')
+            return None
+
+        _logger.debug('successfully authenticated webhook')
+        _logger.debug('POST message: %s', data)
+
+        # Notify the frontend that we received a message from Mercado Pago
+        request.env['bus.bus']._sendone(pos_session_sudo._get_bus_channel_name(), 'MERCADO_PAGO_LATEST_MESSAGE', {})
+
+        # Acknowledge Mercado Pago message
+        return request.make_json_response('[success]')
+
+    def validate_external_reference(self, ext_ref):
+        """
+        An attacker can be aware of the 'external_reference' field existence and can corrupt it,
+        hence we test the format
+        """
+        # parts = ext_ref.split('_')
+        # if len(parts) == 2 and parts[0].isdigit() and parts[1].isdigit():
+        #     return True
+        # return False
+        try:
+            left, right = ext_ref.split('_')
+            return left.isdigit() and right.isdigit()
+        except ...:
+            return False
```

Comment:
Check with your collegue @mart-e , my original code was
```
parts = ext_ref.split('_')
if len(parts) == 2 and parts[0].isdigit() and parts[1].isdigit():
    return True
return False
```

Reply:
@papyDoctor the `...` was to fill with proper exception ^^ but yes regex works too.
Actually, is this method really useful? I mean, you already do this check in the controller above (the `split` and the two `int()`) so it will crash if not in the expected format anyway.

Reply:
I do like this now
```
if not external_reference or not bool(re.match(r'^\d+_\d+$', external_reference)):
    _logger.debug('POST message received with no or malformed "external_reference" key')
    return http.Response(status=204)
```